src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EVP.cs

```diff
@@ -37,24 +22,61 @@ internal static partial class Crypto
         internal static extern int EvpMdSize(IntPtr md);
 
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_EvpMd5")]
-        internal static extern IntPtr EvpMd5();
+        private static extern IntPtr CryptoNative_EvpMd5();
 
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_EvpSha1")]
-        internal static extern IntPtr EvpSha1();
+        private static extern IntPtr CryptoNative_EvpSha1();
 
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_EvpSha256")]
-        internal static extern IntPtr EvpSha256();
+        private static extern IntPtr CryptoNative_EvpSha256();
 
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_EvpSha384")]
-        internal static extern IntPtr EvpSha384();
+        private static extern IntPtr CryptoNative_EvpSha384();
 
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_EvpSha512")]
-        internal static extern IntPtr EvpSha512();
-
+        private static extern IntPtr CryptoNative_EvpSha512();
 
         [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_GetMaxMdSize")]
         private static extern int GetMaxMdSize();
 
+        // For these shared algorithm handles the native functions are idempotent,
+        // so it doesn't matter if there's a parallel race to set the field initially.
+        internal static IntPtr EvpMd5() =>
+            s_md5 != IntPtr.Zero ? s_md5 : (s_md5 = CryptoNative_EvpMd5());
+
+        internal static IntPtr EvpSha1() =>
+            s_sha1 != IntPtr.Zero ? s_sha1 : (s_sha1 = CryptoNative_EvpSha1());
+
+        internal static IntPtr EvpSha256() =>
+            s_sha256 != IntPtr.Zero ? s_sha256 : (s_sha256 = CryptoNative_EvpSha256());
+
+        internal static IntPtr EvpSha384() =>
+            s_sha384 != IntPtr.Zero ? s_sha384 : (s_sha384 = CryptoNative_EvpSha384());
+
+        internal static IntPtr EvpSha512() =>
+            s_sha512 != IntPtr.Zero ? s_sha512 : (s_sha512 = CryptoNative_EvpSha512());
+
+        internal static IntPtr GetDigestAlgorithm(string algorithmName)
```

Comment:
👍 Would be nice to replace hash dispenser with this so we can eliminate the duplication, or open an issue to do it in a follow up.

https://github.com/dotnet/runtime/blob/8b391f60276136954a17fb3dc1f254fb98d82022/src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Unix.cs#L32-L39